enhancement(sinks): Batch buffer rework

bruceg opened the pull request:
Closes #2768 
Closes #2826 

This is a draft for initial review. I still need to ensure this is properly tested, but all the required pieces should be here. There are two bits commented in the code:

1. Some of the old code references "items", and some references "events". Which makes more sense to standardize on for this interface, both internally and externally?

2. The logic in `BatchSink::start_send` and `PartitionBatchSink::start_send` is a little subtle and needs careful attention to make sure I didn't create a new problem.

Commit:
Add new VecBuffer wrapper around Vec

This new buffer type keeps track of its maximum size, and pre-allocates
the buffer.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add BatchSettings parameter to MetricsBuffer

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add BatchSettings parameter to Buffer

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add BatchSettings parameter to JsonArrayBuffer

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Move detection of batch "fullness" to trait Batch

This also drops fn len, as it is no longer necessary.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Simplify BatchSink::linger_elapsed

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Make the Batch::push trait method failable

This requires reworking the logic in BatchSink::start_send, and so
demands some close scrutiny.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Introduce StatefulBuffer wrapper and clean up BatchSink

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Drop superfluous Batch::is_full trait method

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Merge BatchBytesConfig and BatchEventsConfig and adjust sinks

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Account for the separating "," in JsonArrayBuffer

Signed-off-by: Bruce Guenter <bruce@timber.io>

fanatid commented on src/sinks/util/batch.rs:
```diff
@@ -1,47 +1,93 @@
+pub enum BatchError {
+    #[snafu(display("Cannot configure both `max_bytes` and `max_size`"))]
+    BytesAndSize,
+    #[snafu(display("Cannot configure both `max_events` and `max_size`"))]
+    EventsAndSize,
```
```suggestion
    #[snafu(display("Cannot configure both `max_events` and `max_bytes`"))]
    EventsAndBytes,
```

bruceg replied:
No, those two are exactly what they are supposed to be. Due to the backwards compatibility nature of `max_size`, it acts as an alias for either `max_bytes` or `max_events` depending on what `max_size` used to be used for. As such, it is not allowed to configure both at once. It is, on the other hand, possible to configure both `max_bytes` and `max_events`, and that is part of the point of this whole rework.

Now, it might be worth adding a note to that error that `max_size` is now deprecated, but that's different than what you've suggested.

fanatid commented on src/sinks/util/batch.rs:
```diff
@@ -1,47 +1,93 @@
+        events: usize,
+        timeout: u64,
+    ) -> Result<BatchSettings, BatchError> {
+        if bytes == 0 && self.max_bytes.is_some() {
```
`max_size` should be checked too? (as `max_bytes` in `parse_events`)

bruceg replied:
That logic could use some explanatory comment. The `parse_with_events` method is used when the sink used the old `max_size` to configure events. For some of those sinks, it is not possible to set a `max_bytes` and that is indicated by having a default of `0`. The reverse is not true, as all sinks can configure a maximum number of events.

fanatid commented on src/sinks/util/buffer/vec.rs:
```diff
@@ -0,0 +1,49 @@
+        self.batch.is_empty()
+    }
+
+    fn fresh(&self) -> Self {
+        Self {
+            batch: Vec::with_capacity(self.max_events),
+            max_events: self.max_events,
+        }
+    }
```
There few places where code in `new` and `fresh` is same. Maybe it worth add non-pub `new2` function?

For example for `VecBuffer` it will looks like:
```rust
impl<T> VecBuffer<T> {
    pub fn new(settings: BatchSettings) -> Self {
        VecBuffer::new2(settings.events)
    }
    fn new2(max_events: usize) -> Self {
        let batch = Vec::with_capacity(max_events);
        Self { batch, max_events, }
    }
}
impl<T> Batch for VecBuffer<T> {
    fn fresh(&self) -> Self {
        VecBuffer::new2(self.max_events)
    }
}
```

bruceg replied:
Yes, that's worth considering.

lukesteensen commented on src/sinks/util/sink.rs:
```diff
@@ -464,27 +469,33 @@ where
 
                 if let Some(batch) = self.partitions.get_mut(&partition) {
-                    if batch.len() >= self.settings.size {
+                    if !batch.is_empty() {
```
This is a little confusing, but I get that it's in combination with the fullness check above.

bruceg replied:
With the change in the semantics of "full", it's no longer possible to ask a buffer if it is full, only if adding something would overflow, because a buffer that tracks bytes could both be full (can't add the next event) and not full (has more bytes left in the buffer) at the same time. That path leads to insanity. However, since the buffer is fully emptied on send, the old logic of `is_full` is equivalent to `!is_empty`. I'll add an explanatory note.

lukesteensen commented on src/sinks/util/sink.rs:
```diff
@@ -464,27 +469,33 @@ where
-                        return Ok(AsyncSink::Ready);
+                        return match batch.push(item) {
+                            PushResult::Ok => Ok(AsyncSink::Ready),
+                            // FIXME is this the right behavior here and below?
+                            PushResult::Full(item) => Ok(AsyncSink::NotReady(item)),
```
I don't think this is actually reachable since the batch should be empty. But either way it seems like the right behavior. There's no way to get here without calling `poll_complete` and then observing evidence that it did not return `Ready` (i.e. batch being no longer full).

bruceg replied:
That's a relief.

lukesteensen commented on src/sinks/util/sink.rs:
```diff
@@ -464,27 +469,33 @@ where
-                return Ok(AsyncSink::Ready);
+                return match batch.push(item) {
+                    PushResult::Ok => Ok(AsyncSink::Ready),
+                    PushResult::Full(item) => Ok(AsyncSink::NotReady(item)),
```
Here I'm less sure. We should not return `NotReady` without trying to drive things forward via `poll_complete`, and it doesn't seem like we've done that in this branch.

bruceg replied:
Is that just a matter of adding a `self.poll_complete()?` here, or something more?

lukesteensen replied:
The rule is that we shouldn't return `PollReady` unless something below us has returned it. The idea is that goes all the way down to IO or timer operations that register our task to be woken up at the appropriate future time. So we should be doing as much work as possible before we're blocked by `NotReady`.

In this case, that specifically means we should call `poll_complete` and then either `push` again if it returns `Ready` (since that should have cleared us room) or return `NotReady` if it returns the same to us.

It seems to me that some reorganization of this method might be useful. We have a few places with code that seems unreachable and doing this properly is probably going to add some duplication.

bruceg replied:
I tried to do it properly, but couldn't figure any good way without either some duplication, or introducing some enums just for use within this method.

lukesteensen commented on src/sinks/util/sink.rs:
```diff
@@ -493,12 +504,16 @@ where
-        batch.push(item);
-        self.set_linger(partition.clone());
+        match batch.push(item) {
+            PushResult::Full(item) => Ok(AsyncSink::NotReady(item)),
```
Same here I think.

bruceg replied:
This should actually be unreachable unless it's an unusable configuration. The batch was just created fresh the line before, so the push shouldn't report full.

lukesteensen replied:
If that's the case I'd prefer we either panic or reorganize to remove this branch.

bruceg replied:
I set it as `unreachable!`

lukesteensen commented on src/sinks/logdna.rs:
```diff
@@ -64,11 +64,13 @@ pub enum Encoding {
         let sink = BatchedHttpSink::new(
             self.clone(),
-            JsonArrayBuffer::default(),
+            JsonArrayBuffer::new(batch_settings),
             request_settings,
             batch_settings,
```
It's a little funky that we pass the `batch_settings` both to the batch itself and to the sink.

bruceg replied:
I was unhappy with that too. I worked a bit on having `BatchSink::new` etc take a closure to which it would then pass the batch settings to create a new batch, but it was looking like it would add more confusion than it would solve. With the exception of `MetricBuffer` that could also be used to get rid of the `fresh` and `fresh_replace` methods, but I don't see a path to make that work. The only part of `BatchSettings` that the sink actually uses is the timeout, so I could split that part out in the calling convention.

lukesteensen replied:
Yeah, that's interesting. From a user's perspective, size and time limits are both just batch settings, but internally they're handled by two different systems. It could make sense to split the structs accordingly.

bruceg replied:
I went ahead and broke out the size settings to be able to pass them separately.

lukesteensen reviewed (commented):
There seem to be some pitfalls here in the partitioned batching logic. It'd be nice if we could find a way to break apart those two concerns more so we don't have to deal with the compounding complexity, but I realize it'd be a good amount of work to figure that out.

bruceg commented on src/sinks/util/batch.rs:
```diff
@@ -1,47 +1,93 @@
+pub struct BatchConfig {
+    pub max_bytes: Option<usize>,
+    pub max_events: Option<usize>, // FIXME: Should this be "events" or "items"?
     pub max_size: Option<usize>,
```
Yes, that's worth noting.

Commit:
Make note on `max_size` being left for backwards compatibility

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg commented on src/sinks/util/buffer/json.rs:
```diff
@@ -1,36 +1,53 @@
 
     fn fresh(&self) -> Self {
-        JsonArrayBuffer::default()
+        Self {
+            buffer: Vec::new(),
+            total_bytes: 0,
+            settings: self.settings,
+        }
```
Good catch. I think that is a remnant from a previous code pattern.

Commit:
Deduplicate fn fresh in the buffers

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg commented on src/sinks/util/batch.rs:
```diff
@@ -1,47 +1,93 @@
-            size: self.max_size.unwrap_or(size as usize),
-            timeout: Duration::from_secs(self.timeout_secs.unwrap_or(timeout)),
+impl BatchConfig {
+    pub fn parse_bytes(&self, bytes: u64) -> Result<usize, BatchError> {
```
Absolutely, I wasn't that excited either about my old (`unwrap_or`) or new (`parse_bytes`) naming. For those two, `get_bytes_or` sounds good.

bruceg commented on src/sinks/util/batch.rs:
```diff
@@ -1,47 +1,93 @@
-    pub fn unwrap_or(&self, size: u64, timeout: u64) -> BatchSettings {
-        BatchSettings {
-            size: self.max_events.unwrap_or(size as usize),
+    pub fn parse_with_bytes(
+        &self,
+        bytes: u64,
+        events: usize,
+        timeout: u64,
+    ) -> Result<BatchSettings, BatchError> {
```
I thought about that, and like that pattern when there are many parameters. The problem is that using `BatchSettings` will then require setting timeout with `Duration::from_secs` which is extra verbose. I guess the reduced confusion will make it worth it. Any suggestions on the function naming?

lukesteensen replied:
It's a bit more work, but we could consider implementing a builder-style pattern for `BatchSettings` that'd help with that. As far as naming, maybe something like `get_settings_or_default`? Don't love that but it might be a little clearer.

bruceg replied:
I'm not convinced a builder pattern is worth it, particularly since we always specify all three parameters. The suggestion of `get_settings_or_default` is missing the size vs events differentiation for compatibility with `max_size`.

bruceg replied:
Never mind, some sinks don't use `bytes`, so I used a builder pattern after all (with a shortcut) plus a new invention of my own for setting bytes/events from size. I hope this improves clarity.

bruceg commented on src/sinks/util/buffer/metrics.rs:
```diff
@@ -108,83 +110,84 @@ impl Batch for MetricBuffer {
-                    // The resulting Counters could be added up normally
-                    if let Some(MetricEntry(mut existing)) = self.metrics.take(&delta) {
+    fn push(&mut self, item: Self::Input) -> PushResult<Self::Input> {
+        if self.num_items() >= self.max_events {
+            PushResult::Full(item)
```
True. I could rework the logic to push that decision into each of the metric types below if you think it is worth it. It would likely be a pretty small batch size optimization though.

lukesteensen replied:
I don't think it's worthwhile right now. Maybe a later optimization.

bruceg commented on src/sinks/util/sink.rs:
```diff
@@ -212,27 +211,33 @@ where
-        Ok(AsyncSink::Ready)
+        match self.batch.push(item) {
+            PushResult::Ok => Ok(AsyncSink::Ready),
+            PushResult::Full(item) => self.start_send(item),
```
Agreed, I reread that one a number of times to make reassure myself of the same.

bruceg commented:
I just realized an assumption bomb hidden in the logic of this code. Consider trying to add a very large event (in terms of its encoded byte size) to an empty buffer with a smaller bytes limit. The very first push will show as full but leave the buffer empty, which could lead to an infinite loop.

I can think of a couple of ways to handle it, neither of which is entirely satisfactory:
1. Always add the first item to the buffer, even if it would overflow the bytes count. This could lead to oversized batches but would avoid dropping events.
2. Drop this first oversized item with an error but return ok. This too allows for progress but leads to data loss.

Any other ideas?

binarylogic commented:
I'd prefer a configurable option with 2 as the default. This is similar to `buffer.when_full` option. Exceeding the `max_bytes` is likely to cause a 413 from the downstream service anyway, so we should not exceed it unless the user explicitly configures the batch to do so.

Alternatively, we could default to 2 for now and open an issue to add an option later.

Commit:
Rename BatchResult::Full to Overflow

Signed-off-by: Bruce Guenter <bruce@timber.io>

Hoverbear commented:
Agreed, especially if this is meant to be common across many sinks, this way we can tune for the sink (and user)

Commit:
Start work on PushResult::Ok(full)

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Mark a copy of paths in PartitionBatchSink as unreachable

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix handling of overflow in PartitionBatchSink

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Rename BatchConfig::parse_X to get_X_or

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Use a simplified fake-ish builder pattern to build BatchSettings

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Rework how max_size is handled to improve clarity

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Simplify BatchSettings builders with ..self

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Split up batch size and timeout settings

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Run make generate

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
trivial: Mark the BatchSettings builders as const

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Manually track number of bytes written in Buffer

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix aws_s3 sink tests

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Handle individual items that are too large to buffer

If the first event to be added to a buffer is too large, this could
cause a sink pipeline to permanently block. Instead, drop the output
event with an error message but return as if it was added.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Simplify PartitionBatchSink::start_send with a helper

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg commented:
I think this is complete now, all previous issues should be resolved and tests were passing up to the last commit at least.

Commit:
Merge remote-tracking branch 'origin/master' into batch-buffer-rework

Signed-off-by: Bruce Guenter <bruce@timber.io>

Hoverbear commented on src/sinks/util/batch.rs:
```diff
@@ -1,47 +1,137 @@
+pub struct BatchConfig {
+    pub max_bytes: Option<usize>,
+    pub max_events: Option<usize>,
+    // This field is deprecated, left in for backwards compatibility
```
```suggestion
    /// **Deprecated:** This field is deprecated, left in for backwards compatibility
```

bruceg replied:
Done

Hoverbear commented on src/sinks/util/batch.rs:
```diff
@@ -1,47 +1,137 @@
+            ..self
+        }
+    }
+    pub const fn timeout(self, secs: u64) -> Self {
```
It seems like this API could be more flexible if it accepted a `Duration`?

bruceg replied:
It could, but at the cost of writing `Duration::from_secs` in all the batch sinks, and nothing uses anything other than seconds.

Commit:
Fix clickhouse tests

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Reword deprecation comment

Signed-off-by: Bruce Guenter <bruce@timber.io>

lukesteensen reviewed (approved):
Nice! This is a really solid improvement.

Commit:
Merge remote-tracking branch 'origin/master' into batch-buffer-rework

Signed-off-by: Bruce Guenter <bruce@timber.io>

binarylogic commented on .meta/sinks/sematext_logs.toml.erb:
```diff
@@ -25,7 +25,7 @@ write_to_description = "[Sematext][urls.sematext] via the [Elasticsearch API][ur
 <%= render("_partials/fields/_component_options.toml", type: "sink", name: "sematext_logs") %>
 
-<%= render("_partials/fields/_batch_options.toml", namespace: "sinks.sematext_logs.options", common: false, max_events: nil, max_size: 10490000, timeout_secs: 1) %>
+<%= render("_partials/fields/_batch_options.toml", namespace: "sinks.sematext_logs.options", common: false, max_bytes: 10490000, max_events: 100000, timeout_secs: 1) %>
```
I'm curious why `max_events` was set here. Is that limit imposed by Sematext?

binarylogic commented on .meta/sinks/splunk_hec.toml.erb:
```diff
@@ -22,7 +22,7 @@ write_to_description = "a [Splunk's HTTP Event Collector][urls.splunk_hec]"
 <%= render("_partials/fields/_component_options.toml", type: "sink", name: "splunk_hec") %>
 
-<%= render("_partials/fields/_batch_options.toml", namespace: "sinks.splunk_hec.options", common: false, max_events: nil, max_size: 1049000, timeout_secs: 1) %>
+<%= render("_partials/fields/_batch_options.toml", namespace: "sinks.splunk_hec.options", common: false, max_bytes: 1049000, max_events: 10000, timeout_secs: 1) %>
```
Same here. Is this `max_events` count imposed by Splunk?

binarylogic reviewed (approved):
Fantastic work. Really happy this is at the finish line. I left a couple of non-blocking comments. I'm curious why `max_events` was added to some of the sinks. It is just a precaution or are these actually enforced by the downstream services?

bruceg commented:
> I'm curious why `max_events` was added to some of the sinks. It is just a precaution or are these actually enforced by the downstream services?

All sinks now support limiting `max_events` in a batch and as such require a default limit. Neither of the two sinks you asked impose a limit, it's just a number. I derived that number from 1/100th of the byte limit.

binarylogic commented:
I see. We could simply remove the `max_events` limit and only impose the `max_bytes` limit, right? Unless a sink explicitly limits the number of events I don't think we should add it. We should try to align our limits directly with the service.

fanatid commented on src/sinks/util/batch.rs:
```diff
@@ -1,47 +1,138 @@
+            },
+            timeout: self
+                .timeout_secs
+                .map(|secs| Duration::from_secs(secs))
```
```suggestion
                .map(Duration::from_secs)
```

bruceg replied:
Got it, thanks.

bruceg commented:
I could set the default for `max_events`, for those sinks that didn't have it set, to `usize::max_value()` and call it unlimited. That would be the simplest to deal with in the code.

Commit:
Simplify timeout to duration conversion

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Drop all the arbitrary max_events settings

Signed-off-by: Bruce Guenter <bruce@timber.io>

Hoverbear commented:
I'm getting a couple local test failures:

![image](https://user-images.githubusercontent.com/130903/85901595-050f2c00-b7b7-11ea-96f4-76649b4114f4.png)
![image](https://user-images.githubusercontent.com/130903/85901634-15270b80-b7b7-11ea-8b0d-e23739fde9a5.png)

Commit:
Fix up JsonArrayBuffer test for accounting for separating commas

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Drop test batch_sink_allows_the_final_item_to_exceed_the_buffer_size

We explicitly do not want batches to exceed the buffer size.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix up a couple other miscellaneous tests

Signed-off-by: Bruce Guenter <bruce@timber.io>

binarylogic commented on .meta/sinks/aws_cloudwatch_logs.toml.erb:
```diff
@@ -28,7 +28,7 @@ write_to_description = "[Amazon Web Service's CloudWatch Logs service][urls.aws_
 <%= render("_partials/fields/_component_options.toml", type: "sink", name: "aws_cloudwatch_logs") %>
 
-<%= render("_partials/fields/_batch_options.toml", namespace: "sinks.aws_cloudwatch_logs.options", common: false, max_events: 1000, max_size: nil, timeout_secs: 1) %>
+<%= render("_partials/fields/_batch_options.toml", namespace: "sinks.aws_cloudwatch_logs.options", common: false, max_bytes: nil, max_events: 1000, timeout_secs: 1) %>
```
Here, the limits should be:

```
max_bytes: 1,048,576, max_events: 10000
```

As [described in the docs](https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html). This is the most important sink for us and the one that originally raised the issue with our batch limits.

bruceg replied:
Hmm, I missed this in the rest of the noise. Unfortunately, handling this adds some complication, and I think this PR is otherwise almost ready to merge. All of this work is a necessary prelude to handle this case. Can I drop the "closes" for that issue, merge this PR, and finish that work in a separate PR with this as a baseline to avoid dragging this one out further?

binarylogic replied:
That works for me.
